tests/pamiq_core/threads/test_base.py

```diff
@@ -120,3 +128,165 @@ def on_tick(self) -> None:
+        """Test that the Thread class can NOT be initialized with a 'control'
+        THREAD_TYPE attribute."""
+
+        class TestThreadWithControlThreadType(BackgroundThread):
+            THREAD_TYPE = ThreadTypes.CONTROL
+
+            def __init__(self) -> None:
+                super().__init__()
+
+        with pytest.raises(
+            ValueError,
+            match="BackgroundThread cannot be of type 'control'.",
+        ):
+            _ = TestThreadWithControlThreadType()
+
+    def test_thread_status(self, background_thread) -> None:
+        """Test that the thread status returns ReadOnlyThreadStatus
+        instance."""
+        assert isinstance(background_thread.thread_status, ReadOnlyThreadStatus)
+
+    def test_attach_controller(self, background_thread, read_only_controller) -> None:
+        """Test that the controller can be attached to the thread."""
+
+        background_thread.attach_controller(read_only_controller)
+        assert isinstance(
+            background_thread._controller_command_handler, ControllerCommandHandler
+        )
+
+    def test_on_paused(self, background_thread, mocker) -> None:
+        """Test that the on_paused method changes thread status."""
+        background_thread.on_paused()
+
+        assert background_thread.thread_status.is_pause() is True
+
+    def test_on_resumed(self, background_thread, mocker) -> None:
+        """Test that the on_resumed method changes thread status."""
+
+        background_thread.on_resumed()
+
+        assert background_thread.thread_status.is_resume() is True
+
+    def test_start(self, background_thread, mocker) -> None:
+        """Test that the start method actually starts the thread."""
+        spy_thread_start = mocker.spy(background_thread._thread, "start")
+        _ = mocker.patch.object(
+            background_thread, "is_running", return_value=False
+        )  # make this to finish the `run()` method
+
+        background_thread.start()
+
+        spy_thread_start.assert_called_once_with()
+
+    def test_join(self, background_thread, mocker) -> None:
+        """Test that the join method actually joins the thread."""
+        spy_thread_join = mocker.spy(background_thread._thread, "join")
+
+        background_thread.start()
+        background_thread.join()
+
+        spy_thread_join.assert_called_once_with()
+
+    def test_is_alive_when_alive(self, background_thread, mocker) -> None:
+        """Test that the is_alive returns True when the thread is alive."""
+        mocker.patch.object(background_thread._thread, "is_alive", return_value=True)
+
+        assert background_thread.is_alive() is True
+
+    def test_is_alive_when_not_alive(self, background_thread, mocker) -> None:
+        """Test that the is_alive returns False when the thread is not
+        alive."""
+        mocker.patch.object(background_thread._thread, "is_alive", return_value=False)
+
+        assert background_thread.is_alive() is False
```

Comment:
ここは個別にメソッドをテストするよりも、`start`, `join`, `is_alive`をすべてまとめてテストする方が良いです。あと `mocker.spy`で`run`にスパイを潜り込ませるのが良いかと

```suggestion
    def test_thread_life_cycle(self,thread_controller, background_thread, mocker) -> None:
        spy_run = mocker.spy(background_thread, "run")
        assert background_thread.is_alive() is False
        timer = threading.Timer(0.1, thread_controller.shutdown)
        start = time.perf_counter()
        background_thread.start()
        timer.start()
        assert background_thread.is_alive()
        
        background_thread.join()
        assert 0.1 < time.perf_counter() - start < 0.2
        assert background_thread.is_alive() is False
        spy_run.assert_called_once_with()
```

Reply:
なるほど！

こちらの方針で修正してみます！